nitro-protocol/contracts/VirtualMarginApp.sol

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+pragma experimental ABIEncoderV2;
+
+import './interfaces/IForceMoveApp.sol';
+import './libraries/NitroUtils.sol';
+import './interfaces/INitroTypes.sol';
+import {ExitFormat as Outcome} from '@statechannels/exit-format/contracts/ExitFormat.sol';
+
+// NOTE: Attack:
+// Bob can submit a convenient candidate, when Alice in trouble (Way back machine attack)
+
+// Possible solutions:
+// 1: Alice does checkpoint periodically
+// 2: Alice hire a WatchTower, which replicates Alice's states,
+// and challenge in the case of challenge event and missing heartbeat
+
+/**
+ * @dev The VirtualMarginApp contract complies with the ForceMoveApp interface and allows payments to be made virtually from Initiator to Receiver (participants[0] to participants[n+1], where n is the number of intermediaries).
+ */
+contract VirtualMarginApp is IForceMoveApp {
+    struct MarginProof {
+        INitroTypes.Signature leaderSignature;
+        INitroTypes.Signature receiverSignature;
+        uint256 version;
+    }
+
+    enum AllocationIndices {
+        Initiator,
+        Receiver
+    }
+
+    /**
+     * @notice Encodes application-specific rules for a particular ForceMove-compliant state channel.
+     * @dev Encodes application-specific rules for a particular ForceMove-compliant state channel.
+     * @param fixedPart Fixed part of the state channel.
+     * @param proof Array of recovered variable parts which constitutes a support proof for the candidate.
+     * @param candidate Recovered variable part the proof was supplied for.
+     */
+    function requireStateSupported(
+        FixedPart calldata fixedPart,
+        RecoveredVariablePart[] calldata proof,
+        RecoveredVariablePart calldata candidate
+    ) external pure override {
+        // This channel has only 4 states which can be supported:
```

Comment:
@mod This is super interesting. Thank you for putting together the PR. Do you see this contract evolving to be (roughly) similar to the outline [here](https://www.notion.so/statechannels/RFC-27-Nitro-Clearinghouse-App-ce88e4e127114d1abe5d7e6b19b502fe#a9c048e42af143719e10fdbfd7765a4f)? Or is there a different vision for how a Margin Virtual Channel app will function?

Reply:
This seems to be a bit of out date (base on call 12th Jan).